### PR TITLE
Also initialize non-joint components

### PR DESCRIPTION
**Pull request by fmauch**
Before, only joints were reading the initial value parameter. However, also sensors and gpios can have configured initial values.

Initialization of vectors were [called](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/src/mock_components/generic_system.cpp#L191), but the [implementation](https://github.com/ros-controls/ros2_control/blob/88c74ae09f9d067542c84f8ec9bd056e7c4a38dc/hardware_interface/src/mock_components/generic_system.cpp#L428-L461) only checked for joints. This leaves all sensor state_interfaces initialized with `NaN`, which in turn breaks other things such as https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/issues/390.

With the changes suggested here, also sensors and gpios get initialized with the initial values read from the URDF.

I think in conjunction it would make sense to extend the ros2_control_demos with inital values. I've used those as a simplified environment also to verify that my URDF was not the culprit here.

I am not 100% sure whether this would be the most suitable modification, as I haven't completely understood the dataflow here, but that change seemed rather obvious to me. Feel free to point me into another direction if needed.

Anyway, this should also be ported back to Galactic. Probably a direct backmerge is not possible, since the fake->mock renaming didn't happen there, but it should be easy to do nonetheless.

### CODE REVIEW

**Review by destogl — COMMENTED** [review]
Thanks! This is great!

Excellent idea with the example in `ros2_control_demos`.

Could you also add a test for this in [this file](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/test/mock_components/test_generic_system.cpp)? It should be straightforward. One should add another `xml` example and then check if parsing is good. There is already test for other initial values. Extending that test would also be fine.

**Comment by fmauch** [comment]
> Could you also add a test for this in [this file](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/test/mock_components/test_generic_system.cpp)? It should be straightforward. One should add another `xml` example and then check if parsing is good.

sure!

File: hardware_interface/include/mock_components/generic_system.hpp [hunk 1]
```diff
@@ -99,7 +99,8 @@ class HARDWARE_INTERFACE_PUBLIC GenericSystem : public hardware_interface::Syste
 
   void initialize_storage_vectors(
     std::vector<std::vector<double>> & commands, std::vector<std::vector<double>> & states,
-    const std::vector<std::string> & interfaces);
+    const std::vector<std::string> & interfaces,
+    const std::vector<hardware_interface::ComponentInfo> & components);
```
[the remark this answers is not in the document]
Comment (fmauch) [reply hardware_interface/include/mock_components/generic_system.hpp]: resolved in [7da083f](https://github.com/ros-controls/ros2_control/pull/822/commits/7da083f1adc0023743bcaa403a7ec846037829c0)

**Comment by fmauch** [comment]
Oh, I was not aware, that by re-requesting a review I would remove other potential reviewers. Sorry if I destroyed the workflow there.

**Comment by destogl** [comment]
>

**Comment by destogl** [comment]
> Oh, I was not aware, that by re-requesting a review I would remove other potential reviewers. Sorry if I destroyed the workflow there.

No this shouldn't be happening. It's not your fault!

**Comment by destogl** [comment]
FYI: I just converted it to draft and back to re-trigger reviewer-lottery workflow. (closing was unintentional :D)

**Review by destogl — APPROVED** [review]
This looks great! Thanks!

**Review by destogl — DISMISSED** [review]
Thanks for following up!

**Review by destogl — APPROVED** [review]

**Comment by fmauch** [comment]
Any updates on merging this? The failing tests seem to be unrelated but I might be wrong there.

**Review by bmagyar — APPROVED** [review]

**Comment by destogl** [comment]
@mergifyio backport humble

**Pull request merged**